### PR TITLE
Fixed #28298 -- Prevented a primary key alteration from adding a foreign key constraint if db_constraint=False.

### PR DESCRIPTION
**Pull request by jschneier**
https://code.djangoproject.com/ticket/28298

### CODE REVIEW

**Comment by jschneier** [comment]
Any pointers on why the tests failed on mysql but not mysql_gis would be much appreciated. Don't have mysql locally but can install I suppose.

File: tests/migrations/test_commands.py [hunk 1]
```diff
@@ -463,6 +463,15 @@ def test_sqlmigrate_forwards(self):
...
         )
 
+    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_db_constraint"})
+    def test_dbconstraint_alter_field_migration(self):
+        out = io.StringIO()
+        call_command("sqlmigrate", "migrations", "0002", stdout=out)
+        output = out.getvalue().lower()
+
+        index_add_constraint = output.find('add constraint')
+        self.assertEqual(index_add_constraint, -1)
```
Comment (adamchainz) [inline tests/migrations/test_commands.py]: `self.assertNotIn('add constraint', output)`

File: tests/migrations/test_commands.py [hunk 2]
```diff
@@ -463,6 +463,15 @@ def test_sqlmigrate_forwards(self):
             "Transaction end not found or found before operation description (unique_together)"
         )
 
+    @override_settings(MIGRATION_MODULES={"migrations": "migrations.test_migrations_db_constraint"})
+    def test_dbconstraint_alter_field_migration(self):
+        out = io.StringIO()
+        call_command("sqlmigrate", "migrations", "0002", stdout=out)
```
Comment (adamchainz) [inline tests/migrations/test_commands.py]: use single quoted strings everywhere - will be in code style guide soon as per #8284

File: tests/migrations/test_commands.py (hunk 1 above)
Comment (jschneier) [inline tests/migrations/test_commands.py]: I only wrote it this way for consistency with the other tests that parse the string output.

**Comment by charettes** [comment]
Thanks for your patch! The current `add constraint` assertion seems a bit fragile to me.

We don't directly test the schema layer but adding tests in `test_operations` and asserting the constraint doesn't exist by using database introspection seems more appropriate to me.

**Comment by jschneier** [comment]
That's fine. Do you have any pointers or examples of the db introspection? A quick glance through `test_operations.py` and I wasn't really sure

**Comment by jschneier** [comment]
@charettes oh you meant it literally. I didn't realize there was a `connection.introspection` feature. I just rewrote the test although the assert isn't great.

Thoughts on also limiting to cases when `supports_column_check_constraints = True`?

**Comment by charettes** [comment]
@jschneier I think `supports_foreign_key` and `can_introspect_foreign_keys` are the features you want to pass to `skipUnlessDbFeatures`.

**Comment by jschneier** [comment]
That doesn't skip sqlite though. In my testing I found that it doesn't support constraints anyway.

**Comment by charettes** [comment]
@jschneier I think that SQLite constraint enforcement is determined at compilation time. See [#14024](https://code.djangoproject.com/ticket/14204) for the ticket that added support for them.

**Comment by jschneier** [comment]
I'm wrong anway. `supports_foreign_keys = False` on the `sqlite` backend. Just pushed up the change.

File: tests/migrations/test_operations.py [hunk 3]
```diff
@@ -2240,6 +2248,23 @@ def test_run_python_noop(self):
             operation.database_forwards("test_runpython", editor, project_state, new_state)
             operation.database_backwards("test_runpython", editor, new_state, project_state)
 
+    @skipUnlessDBFeature('supports_foreign_keys', 'can_introspect_foreign_keys')
+    def test_no_dbconstraint_added_during_field_change(self):
+        """
+        #28298
```
Comment (timgraham) [inline tests/migrations/test_operations.py]: As discussed in [Coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style), use a descriptive docstring and reserve ticket references for obscure issues. I think the expected behavior is straightforward enough that a ticket reference isn't needed.

File: tests/migrations/test_operations.py [hunk 4]
```diff
@@ -2240,6 +2248,23 @@ def test_run_python_noop(self):
             operation.database_forwards("test_runpython", editor, project_state, new_state)
             operation.database_backwards("test_runpython", editor, new_state, project_state)
 
+    @skipUnlessDBFeature('supports_foreign_keys', 'can_introspect_foreign_keys')
+    def test_no_dbconstraint_added_during_field_change(self):
```
Comment (timgraham) [inline tests/migrations/test_operations.py]: `db_constraint`?

**Comment by charettes** [comment]
@jschneier sorry about the back and forth here, @timgraham is right that it should live in `tests/schema`. I completely missed it was existing as I was expecting to live in `tests/migrations/test_schema`.

**Comment by timgraham** [comment]
I wrote this test for `schema`. The problem is that it doesn't work as a regression test because `new_field.model._meta.related_objects` is empty (even `old_field.model._meta.related_objects` is also because of the `class Meta: apps = new_apps` stuff in `models.py`), hence the modified line is never executed. Any ideas about how to work around that? I played around with various `new_apps` method calls for a little while but didn't make any progress.
```python
def test_no_db_constraint_added_during_field_change(self):
    """
    When a primary key that's pointed to by a ForeignKey with
    db_constraint=False is altered, a foreign key constraint isn't added.
    """
    with connection.schema_editor() as editor:
        editor.create_model(Author)
        editor.create_model(BookWeak)
    self.assertForeignKeyNotExists(BookWeak, 'author_id', 'schema_author')
    old_field = Author._meta.get_field('id')
    new_field = BigAutoField(primary_key=True)
    new_field.model = Author
    new_field.set_attributes_from_name('id')
    with connection.schema_editor() as editor:
        editor.alter_field(Author, old_field, new_field, strict=True)
    self.assertForeignKeyNotExists(BookWeak, 'author_id', 'schema_author')
```

**Comment by charettes** [comment]
@timgraham not sure from the top of my head. I'll try to have a look this weekend.

**Comment by jschneier** [comment]
I wrote substantially the same test and wasn't sure what I was doing wrong as well.

File: tests/schema/tests.py [hunk 5]
```diff
@@ -318,6 +318,35 @@ def test_fk_db_constraint(self):
...
+    @isolate_apps('schema')
+    @skipUnlessDBFeature('supports_foreign_keys', 'can_introspect_foreign_keys')
+    def test_no_db_constraint_added_during_field_change(self):
+        """
+        When a primary key that's pointed to by a ForeignKey with
+        db_constraint=False is altered, a foreign key constraint isn't added.
+        """
+        class Author(Model):
+            class Meta:
+                app_label = 'schema'
```
Comment (charettes) [inline tests/schema/tests.py]: This is because `tests` will be used for `app_label` if not explicitly specified.

**Pull request merged**